torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -482,18 +482,31 @@ class ReadAfterWriteSyncs : public kir::ExprMutator {
       return;
     }
 
+    // An identical but separate flow of timing for cpasync_wait.
+    //  The insertion and tracking mechanism is the same as RAW
+    //  sync insertion since cp.async only writes smem.
+    // Currently the only interaction which is realized by the
+    //  ordering in this function is that in the case when we need both a
+    //  cpasync wait and a block sync before the same expr, we want
+    //  to place the wait before the block sync, since currently there shouldn't
+    //  be any normal case where we explicitly want the wait after a block sync.
+    if (cpasync_wait_before_.size() > 0 &&
+        cpasync_wait_before_.front() == expr) {
+      cpasync_wait_before_.pop_front();
+      auto last_writes = last_cpasync_writes_.front();
+      last_cpasync_writes_.pop_front();
+
+      auto sync_expr = IrBuilder::create<kir::CpAsyncWait>();
+      insertSyncExpr(last_writes, expr, sync_expr, nullptr);
+    }
```

Comment:
Not technically necessary but suggest to change the code below to `} else if (sync_before_size() ...`. It would be more explicit that the two conditions are mutually exclusive.

Reply:
To clarify a bit, they are not mutually exclusive. Actually, most of the use cases I see require both of them. cp.async.wait doesn't sync the threads, it only makes the current thread wait for completion of async copies issued by the same thread.

Reply:
As noted in the previous PR #1440, the case we have where a async.wait is needed but block sync isn't, is when we manually spill persistent registers into shared mem.

Reply:
Oh, I see. Thought it's mutually exclusive.